File: binary_trees/min_value/min_value.py
from typing import Optional
class Node:
    def __init__(self, val):
        self.val = val
        self. left = None
        self.right = None

    def __str__(self):
        return f"value: {self.val}, left: {self.left}, right: {self.right}"

#      1
#     / \
#    2   0
#   /\   /
#  4  5 -1
a = Node(10)
b = Node(2)
c = Node(0)
d = Node(4)
e = Node(5)
f = Node(-1)

a.left = b
a.right = c
b.left = d
b.right = e
c.left = f

def find_min_value_bfs(root: Optional[Node]) -> int:
    queue = [root]
    results = []

    while(len(queue) > 0):
        current = queue.pop(0)
        results.append(current.val)

        if current.left != None:
            queue.append(current.left)
        if current.right != None:
            queue.append(current.right)
    

    return min(*results)

def find_min_value_dfs(root: Optional[Node]) -> int:
    stack = [root]
    results = []

    while(len(stack) > 0):
        current = stack.pop()
        results.append(current.val)

        if current.left != None:
            stack.append(current.left)
        if current.right != None:
            stack.append(current.right)
    
    return min(*results)


def find_min_value_dfs_recursion(root: Optional[Node]) -> int:
    if root == None:
        return 0
    return min(root.val, find_min_value_dfs_recursion(root.left), find_min_value_dfs_recursion(root.right))


print(find_min_value_bfs(a))
print(find_min_value_dfs(a))
print(find_min_value_dfs_recursion(a))

File: binary_trees/min_value/min_value.js
class Node {
    constructor(val) {
        this.val = val;
        this.left = null;
        this.right = null;
    };
}

const a = new Node(1);
const b = new Node(2);
const c = new Node(0);
const d = new Node(4);
const e = new Node(5);
const f = new Node(-1);

//      1
//     / \
//    2   0
//   /\   /
//  4  5 -1
a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.left = f;

function findMinValueBFS(root) {
    let smallest = Infinity;
    const queue = [ root ];

    while(queue.length > 0) {
        let current = queue.shift();
        if(current.val < smallest) smallest = current.val;

        if(current.left) queue.push(current.left);
        if(current.right) queue.push(current.right);
    }

    return smallest;
}

function findMinValueDFS(root) {
    let smallest = Infinity;
    const stack = [root]

    while(stack.length > 0) {
        let current = stack.pop();
        if(current.val < smallest) smallest = current.val;

        if(current.left) stack.push(current.left);
        if(current.right) stack.push(current.right);
    }

    return smallest;
}

function findMinValueDFSRecursion(root) {
    if(root === null) return Infinity;

    return Math.min(root.val, findMinValueDFSRecursion(root.left), findMinValueDFSRecursion(root.right));
}

console.log(findMinValueBFS(a));
console.log(findMinValueDFS(a));
console.log(findMinValueDFSRecursion(a));